cli/src/doc.rs

```diff
@@ -67,59 +70,82 @@ impl DocCommand {
+                            output.to_string_lossy(),
+                            e
+                        )))
+                    })?);
 
-                        let mut has_file_name = false;
+                    Ok((out, Some(output.to_string_lossy())))
+                })
+                .unwrap_or_else(|| {
+                    let docpath = Path::new(DEFAULT_OUT_DIR);
 
-                        if let Some(path) = self.input.files.get(0) {
-                            if let Some(file_stem) = path.file_stem() {
-                                output_file.push(file_stem);
-                                has_file_name = true;
-                            }
-                        }
+                    fs::create_dir_all(docpath).map_err(|e| {
+                        Error::IOError(IOError(format!(
+                            "when creating output path `{}`: {}",
+                            docpath.to_string_lossy(),
+                            e
+                        )))
+                    })?;
+
+                    let mut output_file = docpath.to_path_buf();
+
+                    let mut has_file_name = false;
 
-                        if !has_file_name {
-                            output_file.push("out");
+                    if let Some(path) = self.input.files.get(0) {
+                        if let Some(file_stem) = path.file_stem() {
+                            output_file.push(file_stem);
+                            has_file_name = true;
                         }
+                    }
 
-                        output_file.set_extension(self.format.extension());
+                    if !has_file_name {
+                        output_file.push("out");
+                    }
+
+                    output_file.set_extension(self.format.extension());
+
+                    let out =
                         fs::File::create(output_file.clone().into_os_string()).map_err(|e| {
                             Error::IOError(IOError(format!(
                                 "when opening or creating output file `{}`: {}",
                                 output_file.to_string_lossy(),
                                 e
                             )))
-                        })
-                    })?,
-            )
+                        })?;
+
+                    Ok((
+                        Box::new(out),
+                        Some(Cow::Owned(output_file.to_string_lossy().into_owned())),
+                    ))
+                })?
         };
+
         match self.format {
             DocFormat::Json => doc.write_json(&mut out),
             DocFormat::Markdown => doc.write_markdown(&mut out),
+        }?;
+
+        if !self.stdout {
+            eprintln!(
+                "Documentation written to {}",
+                out_path
+                    .as_ref()
+                    .map(Cow::as_ref)
+                    .unwrap_or(DEFAULT_OUT_DIR)
```

Comment:
```suggestion
                    .expect("There should always be an output path")
```
Since we're always producing an output path if `self.stdout` is not set, I feel it would be better to panic here if no output path gets set. Just as an insurance policy against the path logic changing in the future.

Reply:
Good point, but then I think I prefer to destructure the option instead of relying on the related-but-the-compiler-can't-prove-it boolean flag. That is, if `out_path` is Some, we print it, otherwise we don't print anything.